bdk_chain/src/keychain/keychain_txout_index.rs

```diff
@@ -231,27 +231,25 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
             .script_pubkey();
 
         let index = (keychain.clone(), next_derivation_index);
-        self.inner.add_spk(index.clone(), new_spk);
-        let new_spk = self
-            .inner
-            .script_pubkeys()
-            .get(&index)
-            .expect("we just added it");
+        self.inner.add_spk(index.clone(), new_spk.clone());
         (next_derivation_index, new_spk)
     }
 
-    /// Gets the next usued script pubkey in the keychain i.e. the script pubkey with the lowest index that has not been used yet.
+    /// Gets the next unused script pubkey in the keychain i.e. the script pubkey with the lowest index that has not been used yet.
     ///
     /// ## Panics
     ///
     /// Panics if `keychain` has never been added to the index
-    pub fn derive_next_unused(&mut self, keychain: &K) -> (u32, &Script) {
-        let need_new = self.keychain_unused(keychain).next().is_none();
-        // this rather strange branch is needed because of some lifetime issues
-        if need_new {
-            self.derive_new(keychain)
+    pub fn next_unused<'a>(&'a mut self, keychain: &K) -> (u32, Script) {
+        let existing = {
+            self.keychain_unused(keychain)
+                .next()
+                .map(|(i, s)| (i, s.clone()))
+        };
```

Comment:
> I didn't get it.. Which extra brackets?

@rajarshimaitra 

```suggestion
        let existing = self.keychain_unused(keychain)
            .next()
            .map(|(i, s)| (i, s.clone()));
```

Reply:
Ah right.. I used that to solve that lifetime issue to remove the immutable ref first out of scope,  before taking another mutable ref.. But we don't that anymore as we are cloning data out..

Reply:
Updated in https://github.com/LLFourn/bdk_core_staging/pull/117/commits/a47909ca82ca7b240b32cdfb9a391a5c3a83d722